src/stats.cc

```diff
@@ -0,0 +1,182 @@
+// Copyright 2013, 2014, 2015, 2016, 2017 Lovell Fuller and contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <numeric>
+#include <vector>
+#include <iostream>
+
+#include <node.h>
+#include <nan.h>
+#include <vips/vips8>
+
+#include "common.h"
+#include "stats.h"
+
+class StatsWorker : public Nan::AsyncWorker {
+ public:
+  StatsWorker(
+    Nan::Callback *callback, StatsBaton *baton, Nan::Callback *debuglog,
+    std::vector<v8::Local<v8::Object>> const buffersToPersist) :
+    Nan::AsyncWorker(callback), baton(baton), debuglog(debuglog),
+    buffersToPersist(buffersToPersist) {
+    // Protect Buffer objects from GC, keyed on index
+    std::accumulate(buffersToPersist.begin(), buffersToPersist.end(), 0,
+      [this](uint32_t index, v8::Local<v8::Object> const buffer) -> uint32_t {
+        SaveToPersistent(index, buffer);
+        return index + 1;
+      });
+  }
+  ~StatsWorker() {}
+
+  const int STAT_MIN_INDEX = 0;
+  const int STAT_MAX_INDEX = 1;
+  const int STAT_SUM_INDEX = 2;
+  const int STAT_SQ_SUM_INDEX = 3;
+  const int STAT_MEAN_INDEX = 4;
+  const int STAT_STDEV_INDEX = 5;
+  const int STAT_MINX_INDEX = 6;
+  const int STAT_MINY_INDEX = 7;
+  const int STAT_MAXX_INDEX = 8;
+  const int STAT_MAXY_INDEX = 9;
+
+  void Execute() {
+    // Decrement queued task counter
+    g_atomic_int_dec_and_test(&sharp::counterQueue);
+    using Nan::New;
+    using Nan::Set;
+    using sharp::MaximumImageAlpha;
+
+    vips::VImage image;
+    vips::VImage stats;
+    sharp::ImageType imageType = sharp::ImageType::UNKNOWN;
+
+    try {
+      std::tie(image, imageType) = OpenInput(baton->input, VIPS_ACCESS_SEQUENTIAL);
```

Comment:
Probably best to use the configured `baton->accessMode` instead of `VIPS_ACCESS_SEQUENTIAL` as this defaults to random, which avoids over-computation from the source decoder (and should fix those libtiff test errors).